### DISCUSSION TITLE
Best practices when using relationships?

### DISCUSSION THREAD

**Discussion by ccrvlh** [sqlalchemy/sqlalchemy#8081] [Usage Questions] [answered]
Hi all.

We've been having a hard time dealing with relationships. We currently have somewhere around 60-70 models, and they are all fairly connected. What seems to be happening is that each query makes tens (if not hundreds) of requests to the database.

It seems that if we have:
`Users -> (relationship) Tenants -> (relationship) Dimensions`
and we query `Users`, relationship loading is propagated, so we end up loading a lot of data (up until the `dimensions` relationships) and having really slow queries.

This is most definitely a design problem on our part, and I'm trying to figure out how to deal with it. One thing that came to mind, is to have all relationships with `noload`, and then chose specifically what to load on the query itself. What caught my attention though, is that the docs mention this is a rarely used method, and I wonder why, and if I'm missing something. I would imagine that having the default loading technique for a relationship will load everything always, so I would expected for things to be usually the other way around: you only load the needed relationships.

Maybe the most usual way of dealing with relationships is to lazy load. That would be somewhat of a challenge for us currently, since we parse the models to Pydantic's Schemas, which already have the relationship attributes, so if I'm not mistaken using Pydantic is forcing the attributes to be loaded, I guess it would probably be better to have another layer between SQLAlchemy and Pydantic, have the lazy loading technique in place, and dynamically adding the attribute to Pydantic's model (or not) maybe?

I'm not certain if I'm making the right trade-off analysis here though, and if I'm missing something. I'm mostly used to slightly smaller databases with less data, so that has never been an issue up until now. Any thoughts/experiences on best practices and how to deal with relationships in a way to avoid loading unnecessary data and stop the relationship loading propagation?

Thanks!

**Comment by CaselIT** [comment]
> One thing that came to mind, is to have all relationships with `noload`, and then chose specifically what to load on the query itself.

This the default behavior of the relationship. They aren't loaded in the query, unless a specific loading option is specified in the options.
In the default case they are loaded upon first access on the attribute of an object. 
Have you changed the default option in the relationship?

> Maybe the most usual way of dealing with relationships is to lazy load. That would be somewhat of a challenge for us currently, since we parse the models to Pydantic's Schemas, which already have the relationship attributes, so if I'm not mistaken using Pydantic is forcing the attributes to be loaded, I guess it would probably be better to have another layer between SQLAlchemy and Pydantic, have the lazy loading technique in place, and dynamically adding the attribute to Pydantic's model (or not) maybe

I'm not familiar with pydantic, but if it indeed insists on accessing every attribute even if they are not present on the class, then I guess that may be an use case of noload.

**Reply by ccrvlh** [reply]
We've been trying a few things, so I guess you're right and we unintentionally changed the default behavior. 
As Pydantic parses the model (that will later be parsed as JSON) it does insist on accessing every attribute, which seems to be our problem indeed.
Thanks a lot!

**Reply by CaselIT** [reply]
that pydantic behavior seems something that may be configurable, but I'm not an expert, so I don't know it that's actually the case

**Reply by ccrvlh** [reply]
Yeah, also not sure. The `from_orm` method doesn't seem to be very advanced in the sense that it knows whether the attribute is a relationship or a plain column, but that's something to investigate indeed.

**Comment by zzzeek** [comment, marked as answer]
> Hi all.
> 
> We've been having a hard time dealing with relationships. We currently have somewhere around 60-70 models, and they are all fairly connected. What seems to be happening is that each query makes tens (if not hundreds) of requests to the database.
> 
> It seems that if we have: `Users -> (relationship) Tenants -> (relationship) Dimensions` and we query `Users`, relationship loading is propagated, so we end up loading a lot of data (up until the `dimensions` relationships) and having really slow queries.
> 
> This is most definitely a design problem on our part, and I'm trying to figure out how to deal with it. One thing that came to mind, is to have all relationships with `noload`, and then chose specifically what to load on the query itself. What caught my attention though, is that the docs mention this is a rarely used method, and I wonder why, and if I'm missing something.

noload makes the relationship return None, or an empty list.   If your code accesses this relationship, it's basically non-functional and is returning the wrong answer.  Which leads to, why is your code accessing this relationship if you don't actually need what it represents?   

if you want to prevent the relationships from loading, use lazy="raise".  that's the best approach if your code has places that are loading relationships in an unwanted way which you'd like to catch.  relationships never load unless your code is explicitly telling them to, either by using an eager loader strategy, or by accessing them on the instance.


 > I would imagine that having the default loading technique for a relationship will load everything always, so I would expected for things to be usually the other way around: you only load the needed relationships.

the default loading technique is that no relationships load, until you access them on individual instances.  if you dont access them, nothing happens.

> 
> Maybe the most usual way of dealing with relationships is to lazy load. That would be somewhat of a challenge for us currently, since we parse the models to Pydantic's Schemas, which already have the relationship attributes, so if I'm not mistaken using Pydantic is forcing the attributes to be loaded, I guess it would probably be better to have another layer between SQLAlchemy and Pydantic, 

that's your actual problem (which I could sort of tell by the way you are describing these relationships).

looking at what pydantic schemas does, it's creating a JSON representation of your model, so as such it has to recurse through all the attributes to create this schema.

If you **do** want these relationships represented in your JSON, then pydantic is doing the right thing and you'd want to organize your relationship loading using appropriate eager loading so that the data gets into your JSON in an efficient way.

if you **dont** want these relationships represented in your JSON, as seems to be implied by wanting to put "noload" on them, then pydantic is doing the wrong thing; it's accessing attributes which you dont want represented, and SQL or not this still would represent unnecessary work if so.  so you want to look into configuring pydantic to not descend into paths that you dont want in your JSON in any case.


> 
> I'm not certain if I'm making the right trade-off analysis here though, and if I'm missing something. I'm mostly used to slightly smaller databases with less data, so that has never been an issue up until now. Any thoughts/experiences on best practices and how to deal with relationships in a way to avoid loading unnecessary data and stop the relationship loading propagation?

above all you need explicit control for serialization schemes as to what they access and what they don't.  The approach of just "load every attribute on the object no matter what" is something I wouldn't be using, if that's what's happening now.  pydantic schemas looks to be an explicit configuration approach, if im reading it correctly, so..you'd want to just configure it not to descend into relationship attributes you dont need.

**Reply by ccrvlh** [reply]
Right on, I think that's exactly our problem. Pydantic has a very convenient `from_orm` method, which is great, but it does take away a lot of control from our hands as is. I'm not sure whether I can configure the `from_orm` behavior to stop descending and accessing relationships, will take a look into that.

I'm thinking that passing the responsibility of choosing what to load from the mapping itself to the query seems to make sense then, it would shield Pydantic (or any other parsing lib) from loading unnecessary attributes, and would make us control what (and when) to load relationships. That way we could probably keep having the related attributes declared in the Pydantic Schemas, without loading unnecessary data. 

I'm curious on how would that behave using different parsing libs (`attrs`, `dataclasses`, `schematics`, `marshamallow` etc). Or maybe we could also have a `to_dict()` method on our `BaseMixing` and ditch the `from_orm` method from Pydantic and the instantiates the schema with a dict.... anyways, will make some tests.

Thanks a lot for the (blazing) fast response as always, much appreciated!

**Comment by ccrvlh** [comment]
For other Pydantic users coming in the future, I ended up adding `lazy='noload'` to most of our relationships, and dealing with relationships on a per-query basis mostly `select(MyTable).options(joinedload(MyTable.myrelationship)).filter(....)` and `select(MyTable).options(subqueryload(MyTable.myrelationship)).filter(....)`.

This shields Pydantic from accessing unwanted attributes, and gives a lot more control over whats being loaded. Queries are cleaner, results are easier to reason about and the performance gain was huge in our case (from 1.5 - 2 sec to 0,3-5s). We also view the greater similarity with raw SQL as a bonus (we have been moving to 2.x syntax).

Will keep exploring, but it seems to be the way to go for us, at least for now.
